synapse/handlers/room.py

```diff
@@ -1217,6 +1218,114 @@ async def filter_evts(events: List[EventBase]) -> List[EventBase]:
         return results
 
 
+class TimestampLookupHandler:
+    def __init__(self, hs: "HomeServer"):
+        self.server_name = hs.hostname
+        self.store = hs.get_datastore()
+        self.state_handler = hs.get_state_handler()
+        self.transport_layer = hs.get_federation_transport_client()
+
+    async def get_event_for_timestamp(
+        self,
+        requester: Requester,
+        room_id: str,
+        timestamp: int,
+        direction: str,
+    ) -> Optional[str]:
+        """Find the closest event to the given timestamp in the given direction.
+        If we can't find an event locally or the event we have locally is next to a gap,
+        it will ask other federated homeservers for an event.
+
+        Args:
+            requester: The user making the request according to the access token
+            room_id: Room to fetch the event from
+            timestamp: The point in time we should navigate from in the given
+                direction to find the closest event.
+            direction: ["f"|"b"] to indicate whether we should navigate forward
+                or backward from the given timestamp to find the closest event.
+
+        Returns:
+            `event_id` closest to the given timestamp in the given direction
+
+        Raises:
+            SynapseError if unable to find any event locally in the given direction
+        """
+
+        local_event_id = await self.store.get_event_for_timestamp(
+            room_id, timestamp, direction
+        )
+
+        logger.debug(
+            "get_event_for_timestamp: locally, we found event_id=%s closest to timestamp=%s",
+            local_event_id,
+            timestamp,
+        )
+
+        # If we found an extremity, we should probably ask another homeserver
+        # first about more history in between
+        is_extremity = False
+        if local_event_id:
+            is_extremity = await self.store.check_if_event_is_extremity(
+                room_id, local_event_id
+            )
+        if not local_event_id or is_extremity:
```

Comment:
Do we still want to ask a remote homeserver if the extremity's direction matches `direction`?

eg. if a homeserver is up to date in a room, and we're looking for an event after `timestamp`, and that event happens to be the latest one, we'll end up querying other homeservers fruitlessly.

Reply:
We could skip asking other federated homeservers if the event matches `event_forward_extremities`.

Can change the concept to look for gaps instead of extremities.

Reply:
I'm not sure I explained it well.

Let's say that the local homeserver has 4 events, A, B, C and D, linked linearly. The events before A and after D are not known locally.

Requests 1, 2 and 3 come in, with the timestamp represented by a `|` and direction represented by an arrow:
```
          A    B    C    D                    other events
       <--x----x----x----x-->                  <--x--...

                                      <-| request 1
       request 2 |->
            request 3 |->

           <-| request 4
```

Requests 1 and 2 do the right thing currently.
For request 1, the homeserver finds event D, whose successors are not known locally, so it would ask over federation. There's a good chance other servers know of events closer to the timestamp so this is good.
Request 2 finds event C and no federation requests are done, which is also sensible. Event C may not actually be the closest, but it's hoped this is unlikely.

Request 3 finds event D. The current code would make a federation request, but I'm suggesting that it shouldn't, for the same reason request 2 doesn't make one: the section of history between request 3's timestamp and D is known
Request 4 is basically the same as 3, but mirrored to be backwards and the same reasoning should apply.

In summary, if we are looking backwards and we find a forward extremity*, then we should continue to ask over federation.
If we are looking forwards and we find a forward extremity*, then I suggest we return it.
and the same for backward extremities, with the direction flipped.

*an event whose successors aren't known, not the definition in the docs.

Reply:
Oh, I see the code's been changed so that request 1 wouldn't federate if D is in the `event_forward_extremities` table. That looks very sensible!

Reply:
> In summary, if we are looking backwards and we find an event with no successors, then we should continue to ask over federation.
> If we are looking forwards and we find an event with no successors, then I suggest we return it.
and the same for events with no predecessors, with the direction flipped.

Updated 👍 and split the logic into `is_event_next_to_backward_gap` and `is_event_next_to_forward_gap`